flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1980,6 +2065,97 @@ mlir::Value IntrinsicLibrary::genIOr(mlir::Type resultType,
   return builder.create<mlir::OrOp>(loc, args[0], args[1]);
 }
 
+// ISHFT
+mlir::Value IntrinsicLibrary::genIshft(mlir::Type resultType,
+                                       llvm::ArrayRef<mlir::Value> args) {
+  // A conformant ISHFT(I,SHIFT) call satisfies:
+  //     abs(SHIFT) <= BIT_SIZE(I)
+  // Return:  abs(SHIFT) >= BIT_SIZE(I)
+  //              ? 0
+  //              : SHIFT < 0
+  //                    ? I >> abs(SHIFT)
+  //                    : I << abs(SHIFT)
+  assert(args.size() == 2);
+  auto bitSize = builder.createIntegerConstant(
+      loc, resultType, resultType.cast<mlir::IntegerType>().getWidth());
+  auto zero = builder.createIntegerConstant(loc, resultType, 0);
+  auto shift = builder.createConvert(loc, resultType, args[1]);
+  auto absShift = genAbs(resultType, {shift});
+  auto left = builder.create<mlir::ShiftLeftOp>(loc, args[0], absShift);
+  auto right =
+      builder.create<mlir::UnsignedShiftRightOp>(loc, args[0], absShift);
+  auto shiftIsLarge = builder.create<mlir::CmpIOp>(
+      loc, mlir::CmpIPredicate::sge, absShift, bitSize);
+  auto shiftIsNegative =
+      builder.create<mlir::CmpIOp>(loc, mlir::CmpIPredicate::slt, shift, zero);
+  auto sel = builder.create<mlir::SelectOp>(loc, shiftIsNegative, right, left);
+  return builder.create<mlir::SelectOp>(loc, shiftIsLarge, zero, sel);
+}
+
+// ISHFTC
+mlir::Value IntrinsicLibrary::genIshftc(mlir::Type resultType,
+                                        llvm::ArrayRef<mlir::Value> args) {
```

Comment:
`ISHFTC` might be at the limit of what lowering should generate in inlined code, it's quite a few ops, congrats getting it working :) Right now, let's keep it like that, but if we face issues with it, I think we should consider moving it to some runtime or making it a fir or mlir op (or to force its outlining by setting the related `IntrinsicHandler::outline` to `true` in the intrinsic lowering table).

Reply:
Yes, we do have some alternate choices.  One is to move this completely to the runtime library.  Another is to use `-outline-intrinsics` on this to allow some sharing.  It probably doesn't make sense to add an actual operator anywhere for this, because it would only be useful for exactly this one case.  Even a generic circular shift op wouldn't be very useful even if hardware has one, because it would only apply to the subset of cases where `SIZE = BIT_SIZE(I)`.

So yes, I think this should be reasonable at least for now, if not longer.

Reply:
Follow up - An attempt to automatically outline `ishftc` bumps into the
`todo: cannot outline call to intrinsic ishftc with absent optional argument`
issue.  (The SIZE argument is optional.)

Reply:
Thanks for testing, I missed that `ishftc` had an optional argument. I have not decided yet how this should be handled in outlining. Given outlining is not a required feature so far, I did not put more effort there. Let's go with the inlined version for now then.